cache/grpcproxy/grpcproxy.go

```diff
@@ -0,0 +1,316 @@
+		if err != nil {
+			logResponse(r.errorLogger, "Write", err.Error(), item.Hash)
+			return
+		}
+
+		buf := make([]byte, 2*1024*1024)
+		resourceName := fmt.Sprintf("uploads/%s/blobs/%s/%d", uuid.New().String(), item.Hash, item.SizeOnDisk)
+		firstIteration := true
+		for {
+			n, err := item.Rc.Read(buf)
+			if err != nil && err != io.EOF {
+				logResponse(r.errorLogger, "Write", err.Error(), resourceName)
+				err := stream.CloseSend()
+				if err != nil {
+					logResponse(r.errorLogger, "Write", err.Error(), resourceName)
+				}
+				return
+			}
+			if n > 0 {
+				rn := ""
+				if firstIteration {
+					firstIteration = false
+					rn = resourceName
+				}
+				req := &bs.WriteRequest{
+					ResourceName: rn,
+					Data:         buf[:n],
+				}
+				err := stream.Send(req)
+				if err != nil {
+					logResponse(r.errorLogger, "Write", err.Error(), resourceName)
+					return
+				}
+			} else {
+				_, err = stream.CloseAndRecv()
+				if err != nil {
+					logResponse(r.errorLogger, "Write", err.Error(), resourceName)
+					return
+				}
+				logResponse(r.accessLogger, "Write", "Success", resourceName)
+				return
+			}
+		}
+	default:
+		logResponse(r.errorLogger, "Write", "Unexpected kind", item.Kind.String())
+		return
+	}
+}
+
+func (r *remoteGrpcProxyCache) Put(ctx context.Context, kind cache.EntryKind, hash string, logicalSize int64, sizeOnDisk int64, rc io.ReadCloser) {
+	if r.uploadQueue == nil {
+		rc.Close()
+		return
+	}
+
+	item := backendproxy.UploadReq{
+		Hash:        hash,
+		LogicalSize: logicalSize,
+		SizeOnDisk:  sizeOnDisk,
+		Kind:        kind,
+		Rc:          rc,
+	}
+
+	select {
+	case r.uploadQueue <- item:
+	default:
+		r.errorLogger.Printf("too many uploads queued")
+		rc.Close()
+	}
+}
+
+func (r *remoteGrpcProxyCache) Get(ctx context.Context, kind cache.EntryKind, hash string) (io.ReadCloser, int64, error) {
+	switch kind {
+	case cache.RAW:
+		// RAW cache entries are a special case of AC, used when --disable_http_ac_validation
+		// is enabled. We can treat them as AC in this scope
+		fallthrough
+	case cache.AC:
+		digest := pb.Digest{
+			Hash:      hash,
+			SizeBytes: -1,
+		}
+
+		req := &pb.GetActionResultRequest{ActionDigest: &digest}
+
+		res, err := r.clients.ac.GetActionResult(ctx, req)
+		status, ok := status.FromError(err)
+		if ok && status.Code() == codes.NotFound {
+			return nil, -1, nil
+		}
+
+		if err != nil {
+			logResponse(r.errorLogger, "GetActionResult", err.Error(), digest.Hash)
+			return nil, -1, err
+		}
+		data, err := proto.Marshal(res)
+		if err != nil {
+			logResponse(r.errorLogger, "GetActionResult", err.Error(), digest.Hash)
+			return nil, -1, err
+		}
+
+		logResponse(r.accessLogger, "GetActionResult", "Success", digest.Hash)
+		return io.NopCloser(bytes.NewReader(data)), int64(len(data)), nil
+
+	case cache.CAS:
+		// We don't know the size, so send a FetchBlob request first to get the digest
+		// TODO: consider passign the expected blob size to the proxy?
```

Comment:
This has been one of the complicating factors that I have been putting off dealing with in order to add a grpc proxy backend...

The first proxy backends were added to bazel-remote before it had gRPC support, and were designed for bazel's simpler remote cache REST API. In particular, the REST api only specifies the hash but not the size of CAS blobs, whereas the gRPC API uses hash/size digests.

One way to handle this, would be to add an optional size argument to the Proxy functions, pass -1 if unknown, and let the proxy backend use the value if non-negative.

Another, probably more complicated option, would be to try refactoring a little, to add an alternate path for gRPC proxy backends from the gRPC code. While the setup might be more complicated, it could probably be more efficient, because there would be fewer calls (eg FindMissingBlobs, BatchUpdateBlobs and BatchReadBlobs wouldn't require multiple calls to the backend). I am  not sure what this would look like- but I think it's worth analysing before accepting this feature.